src/physics/base/Applicability.hh

```diff
@@ -0,0 +1,68 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Applicability.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <functional>
+#include <tuple>
+
+#include "base/Assert.hh"
+#include "base/OpaqueId.hh"
+#include "base/Types.hh"
+#include "physics/material/Types.hh"
+#include "Types.hh"
+#include "Units.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Range where a model and/or process is valid.
+ *
+ * This class is used for specifying the ranges of applicability for a physics
+ * model or process. The interval is *open* on the lower energy range and
+ * *closed* on the upper energy. So a threshold reaction should have the lower
+ * energy set to the threshold. Models valid to zero energy but have special
+ * "at rest" models should set upper to zero.
+ *
+ * An unset value for "material" means it applies to all materials; however,
+ * the particle ID should always be set.
+ */
+struct Applicability
+{
+    MaterialDefId    material{};
+    ParticleDefId    particle{};
+    units::MevEnergy lower = zero_quantity();
+    units::MevEnergy upper = max_quantity();
```

Comment:
I'm fine with an infinite starting range for any given model, since we will update that accordingly. Nevertheless, I wonder if there is a more failsafe way. We should `REQUIRE/EXPECT` that a particle is within the valid energy range for the called interactor to be executed. Since we'll compare `applicability_kn.lower/upper` with `particle.energy().value()`, any model will go through without issues if we forget to assign these boundaries and the bug may persist for awhile. If we start both at `zero_quantity` or `max_quantity`, we should spot that missing info pretty quickly on a test.

Reply:
My intent is for the `Applicability` class to be used during setup only for the processes -- the idea is that the `PhysicsParams` will assemble a contiguous energy grid corresponding to each model, to ensure that each process has no overlapping (or missing) energy range between the models. That should be enough to catch a missing boundary, right?

Reply:
Oh, so maybe I misinterpreted the code: `Applicability` is a part of the `Model` class, so I understood that verifying the applicability would be a recurrent task.
If there is going to be a one time execution that maps models over energy ranges that also checks for overlaps, and then use that mapping to select processes/models, then yes, it's pretty safe.